### PR TITLE
Fix: Limit the number of generated questions

### LINKED ISSUE #10694: [Bug]: Unexpected Number of Questions Generated When Requesting FAQ Generation
### Bug Description

When attempting to generate a set of FAQ questions and answers from a document, specifying num_questions_per_chunk=1 unexpectedly results in multiple questions being generated, exceeding the specified limit.

Expected Behavior
With num_questions_per_chunk=1, I expect to generate exactly one question per document or document chunk processed.

Actual Behavior
Despite setting num_questions_per_chunk=1, multiple questions (7 in my case) are generated for a single document or document chunk, indicating that the limit is not being respected or that the document is being split in an unexpected manner.

Edit : from my LLM log (Mixtral 8x7b running on Replicate) I can see that the LLM is called 8 times even with num_questions_per_chunk=1, so the problem is probably not with the num_questions_per_chunk parameter but with the fact that the request to generate questions/answers is sent multiple times. Could this be due to how async functions work?

### Version

0.9.45

### Steps to Reproduce

Prepare a document or a list of documents to be processed.
Use the provided code snippet to generate FAQ questions and answers, ensuring num_questions_per_chunk is set to 1.
Observe that the output contains more questions than expected.

```
import asyncio
from llamaindex import RagDatasetGenerator

async def main():
    documents = SimpleDirectoryReader(input_files=["./document.txt"]).load_data()
    QUESTION_GEN_PROMPT = "Here is a text. Your job is to imagine {num_questions_per_chunk} questions that could be asked about this text."
    
    dataset_generator = RagDatasetGenerator.from_documents(
        documents=documents,
        service_context=service_context,
        question_gen_query=QUESTION_GEN_PROMPT,
        show_progress=True,
        num_questions_per_chunk=1,  # Expected to generate one question per chunk
    )
    
    labelled_rag_dataset = await dataset_generator.agenerate_dataset_from_nodes()
    for example in labelled_rag_dataset.examples:
        print("-----")
        print(f"Question: {example.query}")
        print(f"Answer: {example.reference_answer}")

if __name__ == "__main__":
    asyncio.run(main())

```

### Relevant Logs/Tracbacks

_No response_

### PR DESCRIPTION
**Pull request by tuomastik**
# Description

Limit the number of generated questions to avoid cases where more questions are generated than requested by the parameter `num_questions_per_chunk`.

Fixes #10694 and #10089

Added the fix to all the modules mentioning `num_questions_per_chunk` except the following modules that contain deprecated classes that possibly do not have this issue due to their `num` parameter in `_agenerate_dataset` method:
* `llama-index-core/llama_index/core/evaluation/dataset_generation.py`
  * The module implements `QueryResponseDataset` that is deprecated in favor of `LabelledRagDataset`
* `llama-index-legacy/llama_index/legacy/evaluation/dataset_generation.py`
  * The module implements `DatasetGenerator` that is deprecated in favor of `RagDatasetGenerator`

## New Package?

- [ ] Yes
- [X] No

## Version Bump?

- [X] Yes
- [ ] No

## Type of Change

- [X] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

```python
from llama_index.core.llama_dataset.generator import RagDatasetGenerator
from llama_index.llms.ollama import Ollama
from llama_index.core.schema import TextNode

node = TextNode(
    text="New York, often called New York City or simply NYC, "
    "is the most populous city in the United States, located at the "
    "southern tip of New York State on one of the world's largest "
    "natural harbors. The city comprises five boroughs, each of which "
    "is coextensive with a respective county. New York is a global center "
    "of finance and commerce, culture and technology, entertainment and "
    "media, academics and scientific output, and the arts and fashion, and, "
    "as home to the headquarters of the United Nations, is an important "
    "center for international diplomacy. New York City is the center of "
    "the world's principal metropolitan economy."
)

llm = Ollama(model="gemma:2b", request_timeout=60.0)
num_questions_per_chunk = 1
for i in range(10):
    gen = RagDatasetGenerator(
        nodes=[node],
        llm=llm,
        num_questions_per_chunk=num_questions_per_chunk,
    )
    generated_questions = gen.generate_questions_from_nodes()
    print(
        f"Number of questions expected: {num_questions_per_chunk} - "
        f"Number of questions generated: {len(generated_questions.examples)}"
    )
```

## Suggested Checklist:

- [X] I have performed a self-review of my own code
- [X] My changes generate no new warnings
- [X] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-legacy/llama_index/legacy/finetuning/cross_encoders/dataset_gen.py
Comment (nerdai) [inline llama-index-legacy/llama_index/legacy/finetuning/cross_encoders/dataset_gen.py:1]: let's not modify `llama-index-legacy`

File: llama-index-legacy/llama_index/legacy/finetuning/embeddings/common.py
Comment (nerdai) [inline llama-index-legacy/llama_index/legacy/finetuning/embeddings/common.py:1]: let's not modify `llama-index-legacy`

File: llama-index-legacy/llama_index/legacy/llama_dataset/generator.py
Comment (nerdai) [inline llama-index-legacy/llama_index/legacy/llama_dataset/generator.py:1]: let's not modify `llama-index-legacy`

File: llama-index-legacy/pyproject.toml
Comment (nerdai) [inline llama-index-legacy/pyproject.toml:1]: let's not modify `llama-index-legacy`

**Review by nerdai — APPROVED** [review]
Thanks @tuomastik! Looks good. Though let's not update legacy llama-index at this point (I've marked in my review where we can revert your changes).

I guess this doesn't cover the case where less than the desired questions were generated, but this problem existed even before your PR

File: llama-index-legacy/llama_index/legacy/finetuning/embeddings/common.py
Comment (tuomastik) [reply llama-index-legacy/llama_index/legacy/finetuning/embeddings/common.py:1]: ✅ Updated the pull request accordingly

File: llama-index-legacy/pyproject.toml
Comment (tuomastik) [reply llama-index-legacy/pyproject.toml:1]: ✅ Updated the pull request accordingly

**Comment by nerdai** [comment]
@tuomastik that makes sense. Maybe for now, we should at least raise a warning to the user that the LLM call resulted in less than the desired amount of questions per chunk. Similar to here: https://github.com/run-llama/llama_index/blob/8c23e0140ab5b74f1fd8ade22c29b4bd758cec46/llama-index-core/llama_index/core/ingestion/pipeline.py#L713

**Comment by tuomastik** [comment]
> Maybe for now, we should at least raise a warning to the user that the LLM call resulted in less than the desired amount of questions per chunk. 

✅ Done

**Review by nerdai — APPROVED** [review]
Lgtm!

**Review by logan-markewich — APPROVED** [review]

**Pull request merged**